Documentation/designs/agent-deployment-pattern-sidecar.md

```diff
@@ -0,0 +1,116 @@
+
+In addition to telling the deployment model, the Agent CR will be the source of truth for remote-write configuration, such as URL and authentication. A change to the remote-write configuration would still require a hot reload of potentially millions of agent sidecar containers, but by avoiding having the remote-write configuration in pod annotation we at least avoid requiring that the Pod manifest also needs to be upgraded.
+
+If different sets of pods require different remote-write configurations, then multiple PrometheusAgent CRs are needed. This means that the pod also needs to specify which Agent CR will inject the sidecar:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: example
+  annotations:
+    prometheus.operator.io/scrape: "true"
+    prometheus.operator.io/path: "/metrics"
+    prometheus.operator.io/port: "8080"
+    prometheus.operator.io/scrape-interval: "60s"
+    prometheus.operator.io/agent-selector: "monitoring/agent-example"
+spec:
+...
+---
+apiVersion: monitoring.coreos.com/v1alpha1
+kind: PrometheusAgent
+metadata:
+  name: agent-example
+  namespace: monitoring
+  spec:
+    mode: sidecar
+    remoteWrite:
+      - url: https://example.com
+```
+
+With a visualization:
+
+![Sidecar Deployment Pattern](assets/agent-deployment-pattern-sidecar/sidecar-deployment-pattern.png "Sidecar Deployment Pattern")
+
+### What to do with ServiceMonitor, PodMonitor, and Probe selectors?
+
+With the sidecar approach, our goal is to scale Prometheus horizontally while avoiding impact in the Kubernetes API. It wouldn't make sense for a sidecar to also scrape metrics from other pods.
+
+If `mode` is set to `sidecar`, a validating webhook would forbid PrometheusAgent CRs are created/updated with the following fields:
+
+* `serviceMonitorSelector`
+* `serviceMonitorNamespaceSelector`
+* `podMonitorSelector`
+* `podMonitorNamespaceSelector`
+* `probeSelector`
+* `probeNamespaceSelector`
+
+## Caveats
+
+### Config Hot Reload
+
+There is two ways to change Prometheus configuration now, 1) by changing annotation on the pod and 2) by changing the remote-write field in PrometheusAgent CRD. The first one will only trigger a hot reload for the involved pod, but the latter has the pottention to trigger millions of hot reloads, depending on the scale of the cluster. 
+
+While there is no research regarding the config-reloader efficiency, this particular container might become problematic for huge scale environments.
+
+### WAL not optimized for small environments
+
+Prometheus Write-Ahead-log(WAL) is stored as a sequence of numbered files with 128MiB each by default. Which means that, by default, at least 128MiB is needed for running Prometheus Agent if we ignore every other part of Prometheus. Using a sidecar, we're optimizing for horizontal scale and 128MiB might be much more than necessary to store metrics from a single Pod.
+
+### Lack of High-Availability setup
```

Comment:
IMO this is fine. A sidecar is tied to the lifecycle of the pod, and comes and goes. If the agent sidecar is broken, the pod is broken. If the node breaks, all sidecars will break, but so will the application container.

Reply:
Actually the HA of Monitoring for the given application is based on the number of replicas for this application.
I mean if the service is running with 2 replicas you already have Prometheus HA.